SE-0139: NSValue bridging

jckarter opened the pull request:
For every struct type for which the frameworks provides an NSValue category for boxing and unboxing values of that type, provide an `_ObjectiveCBridgeable` conformance in the Swift overlay that bridges that struct to NSValue, allowing the structs to be used naturally with id-as-Any APIs and Cocoa container classes. This is mostly a matter of gyb-ing out boilerplate using `NSValue.init(bytes:objCType:)` to construct the instance, `NSValue.objCType` to check its type when casting, and `NSValue.getValue(_:)` to extract the unboxed value, though there are a number of special snowflake cases that need special accommodation:

- To maintain proper layering, CoreGraphics structs need to be bridged in the Foundation overlay.
- AVFoundation provides the NSValue boxing categories for structs owned by CoreMedia, but it does so using its own internal subclasses of NSValue, and these subclasses do not interop properly with the standard `NSValue` subclasses instantiated by Foundation. To do the right thing, we therefore have to let AVFoundation provide the bridging implementation for the CoreMedia types, and we have to use its category methods to do so.
- SceneKit provides NSValue categories to box and unbox SCNVector3, SCNVector4, and SCNMatrix4; however, the methods it provides do so in an unusual way. SCNVector3 and SCNVector4 are packaged into `CGRect`s and then the CGRect is boxed using `valueWithCGRect:`. SCNMatrix4 is copied into a CATransform3D, which is then boxed using `valueWithCATransform3D:` from CoreAnimation. To be consistent with what SceneKit does, use its category methods for these types as well, and when casting, check the type against the type encoding SceneKit uses rather than the type encoding of the expected type.

jckarter commented:
@swift-ci Please test

jrose-apple commented on lib/AST/ASTContext.cpp:
```diff
@@ -3951,8 +3951,17 @@ bool ASTContext::isStandardLibraryTypeBridgedInFoundation(
           nominal == getAnyHashableDecl() ||
-          // Weird one-off case where CGFloat is bridged to NSNumber.
-          nominal->getName() == Id_CGFloat);
+          // Foundation is a dependency of CoreGraphics, but
```
Foundation isn't actually a dependency of CoreGraphics; IIRC the two frameworks are unrelated in iOS, and related the other way on macOS.

jckarter replied:
Interesting. The Swift overlays are set up in CMake such that Foundation's overlay depends on the CoreGraphics one on all platforms. Would it make sense to change it the other way, if that's purely a Swift-build-system layering and not a system one?

jrose-apple replied:
"Foundation is a dependency of CoreGraphics" means "CoreGraphics depends on Foundation", no?

jrose-apple replied:
This check also strikes me as incredibly general. Should we just give up and say all imported types might be bridged at some point?

jckarter replied:
Sorry, I meant "Foundation depends on CoreGraphics". Treating all imported types as possibly bridged might be prudent for ABI compatibility, if we want to reserve the right to continue adding new bridging relationships after ABI stability is declared.

jrose-apple replied:
I'm less concerned about ABI stability than about us either being too general or not general enough. Either we think it's worth maintaining this list, or we don't.

jrose-apple replied:
(You're right that it won't be possible to add cross-module bridging relationships post-stability, but that seems reasonable.)

jckarter replied:
@atrick or @aschwaighofer, how big of a perf impact do you think it'd be if we assumed all imported C value types might have an `_ObjectiveCBridgeable` conformance?

atrick replied:
What kind of an impact do you have in mind? Generating a bunch of conformances bloats the binary and metadata. I guess that could be measured. I don't think the optimizer makes use of knowledge of the absence of a conformance. I don't think we could make that assumption for a public type.

jckarter replied:
@atrick It looks like we *do* try to make optimization assumptions for the lack of `_ObjectiveCBridgeable` conformances by disallowing conformances to that protocol outside of the type's defining module, except for this hardcoded list of exceptions. I'm not sure how valuable that is anymore in the world of `id`-as-`Any`, since there's at least one way to bridge everything to and from a class type now.

jrose-apple commented on test/IRGen/builtins_objc.swift:
```diff
@@ -0,0 +1,35 @@
+
+
+func getObjCTypeEncoding<T>(_: T) {
+  // CHECK: call void @use({{.* i8]\*}} [[INT32:@[0-9]+]],
```
You should be able to check the actual encode strings too, right?

jckarter replied:
I could, but I wasn't sure that the churn of tracking down and conditionalizing the exact encode strings for each type on each supported platform really gave meaningful coverage over just verifying that the builtin works at all. Clang ought to exercise the underlying type encoding code fairly well already.

jrose-apple replied:
Sure, but nothing *is* checking that the builtin works at all.

jrose-apple commented on stdlib/public/SDK/SceneKit/CMakeLists.txt:
```diff
@@ -14,7 +14,7 @@ add_swift_library(swiftSceneKit ${SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES} IS_SDK_
-  SWIFT_MODULE_DEPENDS_WATCHOS Foundation
-  SWIFT_MODULE_DEPENDS simd
+  SWIFT_MODULE_DEPENDS_WATCHOS
+  SWIFT_MODULE_DEPENDS Foundation simd
```
You can probably just remove the previous line, but it wasn't incorrect the old way—the non-watchOS platforms already indirectly depend on Foundation.

jckarter replied:
Good catch.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@jrose-apple You've already glanced over this, but do you mind reviewing for inclusion in 3.0.1 as well?

jrose-apple commented:
I'll do another pass tonight; the first pass wasn't too thorough.

jrose-apple commented:
Seems good for now. Still a few open questions but nothing that would affect the 3.0 branch.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@jrose-apple I updated the comment in ASTContext.cpp and added some tests for specific type encodings that ought to be stable across platforms. How's this look for 3.0 now?

jrose-apple commented:
Ship it!

jckarter commented:
@swift-ci Please test

jckarter commented:
@phausler Is this Foundation test failure expected, or transient?

```
TestFoundation/TestNSURLSession.swift:40: error: TestURLSession.test_dataTaskWithURL : Asynchronous wait failed - Exceeded timeout of 12.0 seconds, with unfulfilled expectations: data task
11:05:05 TestFoundation/TestNSURLSession.swift:42: error: TestURLSession.test_dataTaskWithURL : XCTAssertEqual failed: ("unknown") is not equal to ("Kathmandu") - test_dataTaskWithURLRequest returned an unexpected result
11:05:05 Test Case 'TestURLSession.test_dataTaskWithURL' failed (12.006 seconds).
```

jckarter commented:
@swift-ci Please test

phausler commented:
yea that failure is transient, we have seen it flutter a bit and it really needs to be replaced with loopback testing

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test
